src/relations/shared_db.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing the implementation of the legacy shared-db relation."""
+
+
+import logging
+
+from charms.mysql.v0.mysql import MySQLCreateApplicationDatabaseAndScopedUserError
+from ops.charm import CharmBase, RelationChangedEvent, RelationDepartedEvent
+from ops.framework import Object
+from ops.model import BlockedStatus
+
+from constants import LEGACY_DB_SHARED, PASSWORD_LENGTH, PEER
+from utils import generate_random_password
+
+logger = logging.getLogger(__name__)
+
+
+class SharedDBRelation(Object):
+    """Legacy `shared-db` relation implementation."""
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "shared-db-handler")
+
+        self._charm = charm
+
+        self.framework.observe(
+            self._charm.on[LEGACY_DB_SHARED].relation_changed, self._on_shared_db_relation_changed
+        )
+        self.framework.observe(
+            self._charm.on[LEGACY_DB_SHARED].relation_departed, self._on_shared_db_departed
+        )
+
+        self.framework.observe(self._charm.on.leader_elected, self._on_leader_elected)
+
+    @property
+    def _peers(self):
+        return self.model.get_relation(PEER)
```

Comment:
self does not have a `model`. did you mean `self.charm.model.get_relation(PEER)`? in fact, i dont think this is being used

Reply:
Actually that's a neat thing. Because `SharedDBRelation` inherits from `ops.framework.Object` it has direct access to `model`. The `_ peers` is being used for caching user password.